### DISCUSSION TITLE
React vs Vue vs Angular vs etc...

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#129] [React]
## SPA

1998년 XHR의 대두로 서버에서 필요한 데이터만을 받아 JS에서 이를 동적으로 HTML에 반영할 수 있게 되었다.(JSON은 2001년에 최초 등장) 2005년 이런 방식이 공식적으로 AJAX라는 이름을 갖게 되고 한페이지 내에서 머무르면서 필요한 데이터를 서버에서 받아와서 필요한 부분만 업데이트 하는 방식이 유행하게 됩니다. 이를 SPA라고 하죠. 하나의 데이터가 바뀌면 모든 페이지가 업데이트 되는 MPA와는 달리 전체 페이지를 업데이트 할 필요가 없으므로 빠르고 깜빡임이 없어집니다. 또한 서버의 템플릿도 클라이언트로 분산되고 필요한 리소스만 요청하므로 성능 또한 좋아집니다.

 이러한 SPA의 유행에 따라 이를 쉽게 구현할 수 있는 프레임워크, 라이브러리들이 등장하는데요. 대표주자가 바로 React, Angular, Vue가 있습니다.

<img width="756" alt="스크린샷 2022-10-23 오후 10 33 21" src="https://user-images.githubusercontent.com/65863017/197395212-3ca9aa15-cd40-420d-aae6-377b6bb3c935.png">

## Angular

 Angular는 2010년 구글에 의해 발표된 AngularJS의 업그레이드 버전입니다. 버전이 2.0으로 업그레이드 되면서 AngularJS와의 하위호환은 전혀 되지 않는 모습이었습니다. 기존의 JS기반에서 TS로 버전이 완전히 바뀌게 되면서 공식문서에서도 Typescript 기준으로 작성된 예시 코드들이 다수입니다. 

 라우팅, 상태 관리, form 유효성과 같은 툴을 모두 AngularJS내에 내재하고 있는 All-in-one 프레임워크라고 말할 수 있습니다.

## Vue

 Vue는 React와 Angular의 특징을 섞어놓은 모습이라고 할 수 있습니다. React의 가상 돔과 Angular의 양방향 바인딩을 모두 제공하고 있는 오픈소스 JS프레임워크입니다. 전 구글 직원인 Even You에 의해 만들어졌고 광범위한 문서와 간단한 문법으로 JS에 익숙하지 않은 초심자들이 많은 선택을 합니다.

 JSX를 권장하는 React와는 다르게 Angular처럼 HTML 마크업 기반의 템플릿 문법을 사용합니다. Angular처럼 내장 모듈이 정해진 것은 아니지만 공식문서에서도 나와있듯이 vuex와 같은 특정 모듈들을 권장하는 모습입니다.

## React

 React는 2013년 5월 페이스북에 의해 발표된 UI를 위한 자바스크립트 라이브입니다. Angular와는 다르게 개발에 필요한 모듈을 기본으로 제공하지 않고 서드파티 라이브러리에 의존하고 그 사용은 유저에게 일임하고 있습니다. 

 컴포넌트 기반의 개발로서 다른 라이브러리와는 다르게 JS에 조금 더 치중되어 있습니다. 다른 프레임워크들이 HTML 마크업 기반으로 템플릿 문법을 추가하는 반면 React는 XML 포맷 템플릿을 직접 자바스크립트에 내장시키는 형태로 JSX 기술을 주력으로 사용하고 있습니다.

 그리고 React는 기본적으로 단방향 데이터 바인딩으로 모델이 변경되면 뷰가 변경되는 방법만을 허용합니다. 양방향 데이터 바인딩을 허용하는 Angular와 Vue와는 다른 모습입니다. 개발 단계에서 코드 작성시간이 더 필요하지만 예측가능성이 높아 디버깅과 오류 검출에 유리합니다.

## Angular vs Vue vs React

그럼 이제 본격적으로 대표적인 3개의 프레임워크를 비교해봅시다.

### 진입장벽

 진입 장벽은 JS를 처음 접해보는 사람의 기준으로서는 Vue.js가 가장 낮다. HTML과 JS 기본에 가깝고 어느정도 정형화된 룰을 제공하고 있어서 해당 룰을 따라가면서 구현하면 쉽게 구현해나갈 수 있다. 하지만 애플리케이션의 크기가 커지고 복잡해질수록 프로젝트 설정 또한 다룰줄 알아야 하고 패턴 또한 복잡해질 수 있다. 

 Angular는 여러 문서들도 TS로 이루어져있고 기본 라우팅, 상태관리, form관리 등을 제공하고 있어서 학습 허들이 높다. 또한 기본 JS와는 많이 다른 문법으로 이루어져 있어 3개의 프레임워크 중 가장 복잡하다고 할 수 있다.

 React는 위의 라이브러리 보다 훨씬 Javascript 친화적이다. JS에 익숙한 사용자라면 vue보다 쉽게 접근할 수 있다. React에서 주로 사용되는 JSX 또한 Javascript 문법이고 ES6구문들과 최적화 되어 있기 때문에 기존 JS 유저는 자유성 높게 사용할 수 있다.

### 성능

 사실 성능적인 측면에서는 크게 다르지 않다. 초기 로딩 속도는 Angular가 라우팅, 상태관리 등을 모두 지원하는 모듈을 함께하고 있기 때문에 뒤쳐지지만 애플리케이션의 크기가 커지면 vue와 react 보다 빠른 퍼포먼스를 보여준다.

### 커뮤니티

 커뮤니티는 압도적으로 react가 큰 모습을 보이고 있다. npm 다운로드 수만 봐도 angular와 vue를 합친것의 배가 넘는다. 또한 자유로운 모듈을 지원하기 때문에 기타 추가적인 커뮤니티 또한 react가 압도하는 모습을 확인할 수 있다.

<img width="1340" alt="스크린샷 2022-10-23 오후 8 53 49" src="https://user-images.githubusercontent.com/65863017/197395116-22d14387-620e-4cb9-b502-0c0d5532d802.png">

### Etc

 **svelte**는 2016년 11월 발표된 오픈 소스 프론트엔드 프레임 워크 입니다. 가상돔을 사용하지 않으며  코드의 양을 줄이는 것을 중요시 합니다. 

 **backbonejs**는 ****2010년 발표된 REST JSON인터페이스 기반 자바스크립트 라이브러리로 MVC패턴에 기반을 둔다. 규모가 커지면 관리가 힘들어집니다.

**solid**는 2021년 발표된 프론트엔드 프레임 워크입니다. JSX 문법을 사용하고 빠른 성능과 정확한 반응성을 강점으로 내세우고 있습니다. 기본적으로 React와 굉장히 유사한 모습을 보이지만 차이점이라고 한다면 속도와 번들 사이즈에서 React보다 우월한 모습을 보이고 있습니다.

### Flexible

React는 vue, angular와는 프레임워크가 아닌 라이브러리라고 주장하고 있습니다. 그만큼 다양한 개발자 접근성을 보장한다는 것이죠. react는 애플리케이션을 위한 컴포넌트를 만드는데 집중합니다. 때문에 모바일에서도 데스크탑 앱에서도 적용이 가능한것입니다.

 또한 이러한 자율성이 많은 생태계를 불러옵니다. 추가적인 전역 상태관리 라이브러리만 해도 redux, recoil, jotai, zustand 등 수많은 생태계가 존재합니다. 이러한 생태계 중에서 개발자가 자신에 맞는 라이브러리를 골라서 사용할 수 있는 유연성을 보장합니다.

### Great DX

 위의 flexible의 연장선으로 react는 자유로운 DX환경을 제공합니다. angular와 vue는 strict하게 자신들이 내놓은 문법을 사용하길 권장합니다.

![image](https://user-images.githubusercontent.com/65863017/197395248-ef45f08b-ac25-496b-bee2-8ad4a6c70bdb.png)

![image](https://user-images.githubusercontent.com/65863017/197395258-a0ac76ad-538d-4911-a06f-65024a134b30.png)

기존의 html에 없는 attribute들이 추가되고 익숙치 않은 템플릿 문법으로 개발자들을 당황하게 만들고는 합니다. React는 위 프레임워크들과는 다르게 기존의 html attribute를 보존하는 형태로 최대한 익숙한 환경의 JSX 문법을 지원하고 있습니다. 따라서 기존에 JS에 익숙한 개발자들은 자유롭게 이를 활용할 수 있는 것이죠.

![image](https://user-images.githubusercontent.com/65863017/197395270-a03f5faf-3a52-4de4-89fb-326f51747713.png)

### Large Community & Jobs

 사실 이점이 React를 사용하는 가장 큰 이유이지 않을까 싶습니다. React는 다른 프레임워크와는 비교도 안될정도로 큰 생태계를 자랑하고 있습니다. 큰 생태계라함은 참고할 수 있는 레퍼런스도 다양해지고 또한 개발자들을 구하는 것도 채용됨에도 훨씬 유리함을 갖는다고 말할 수 있을 것 같습니다.

### 참고자료

[https://youtu.be/iZ9csAfU5Os](https://youtu.be/iZ9csAfU5Os)

[https://www.samsungsds.com/kr/insights/frameworks.html](https://www.samsungsds.com/kr/insights/frameworks.html)

[https://blog.udemy.com/react-js-vs-angular-vs-vue-js-which-is-the-best-javascript-framework/](https://blog.udemy.com/react-js-vs-angular-vs-vue-js-which-is-the-best-javascript-framework/)

[https://www.theteams.kr/teams/371/post/64183](https://www.theteams.kr/teams/371/post/64183)

**Comment by airman5573** [comment]
레종드..

**Comment by airman5573** [comment]
바닐라 vs React

**Comment by airman5573** [comment]
양방향 vs 단방향

**Comment by airman5573** [comment]
왜 리액트는 단방향? 앵귤러 뷰는 양방향?

**Comment by airman5573** [comment]
단방향을 리액크 코드 단에서 설명해주세요

**Comment by byhhh2** [comment]
flux 패턴이란?

**Reply by byhhh2** [reply]
- flux 패턴의 장점은?

**Comment by byhhh2** [comment]
어떻게 react를 접하게 되었나요?

**Comment by byhhh2** [comment]
react가 가장 큰 생태계를 가지게 된 이유는?

**Comment by byhhh2** [comment]
react 학습법

**Reply by byhhh2** [reply]
블로그를 참고해 공부할 때의 단점

**Comment by byhhh2** [comment]
본인이 속한 팀에서 react를 사용하지 않는다면 어떻게 빠르게 학습하실건가요?

**Comment by byhhh2** [comment]
react로 개발하면서 어려운점

**Reply by byhhh2** [reply]
react 테크.. (병민 알려줘)

**Comment by byhhh2** [comment]
react 관련 라이브러리 사용하면서 겪은 어려운점 (react-query, emotion .. )

**Comment by byhhh2** [comment]
팀 프로젝트를 진행하면서 어려운 점?

**Comment by byhhh2** [comment]
API 설계시 가장 어려웠던 점

**Comment by byhhh2** [comment]
컴포넌트에 대해 설명해주세요.

**Reply by byhhh2** [reply]
컴포넌트를 사용할 때의 장점?

**Comment by byhhh2** [comment]
컴포넌트를 설계하는 본인만의 방법

**Comment by airman5573** [comment]
단방향 vs 양방향 조금 더 공부 필요해 보여요!
